src/bar.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use crate::pci_config::{validate_dword_alignment, Error as ConfigSpaceAccessError};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+/// Error type for PCI Base Address Register accesses.
+pub enum Error {
+    /// Invalid access in the configuration space.
+    Access(ConfigSpaceAccessError),
+    /// Memory region defined by the BAR address and lenght is invalid.
+    BarAddressInvalid(u64, u64),
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match self {
+            Access(e) => write!(f, "config space access: {}", e),
+            BarAddressInvalid(addr, len) => {
+                write!(f, "out of bounds BAR addr {} len {}", addr, len)
+            }
+        }
+    }
+}
+
+impl From<ConfigSpaceAccessError> for Error {
+    fn from(e: ConfigSpaceAccessError) -> Self {
+        Error::Access(e)
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Enum representing the "prefetchable" state masks of BARs (bit 3 in the register).
```

Comment:
It think there might be some confusion here ( at least in the way this is represented in Rust). The code below looks like it wants to represent both the mask of the prefetchable bit (0x08), and its value (either 0 or 1). Since a bar can either be prefetchable or non-prefetchable, there is no need to use an enum to represent the value. As for the mask (i.e. 0x08), it can also just be a const define in this file instead.

Enums are typically used to express more complex types that can't otherwise be expressed with bools/Option.